### PR TITLE
support pushing bitwise operators and LogicXor down to kv

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit 
1. splits ./distsql/xeval/eval.go into some small files to make the function of every file more explicit.
2. supports pushing bitwise operators and LogicXor down to kv.
PTAL @hanfei1991 @coocood @shenli @zimulala @tiancaiamao

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
I split eval.go into five small files **to classify different type of operators.**
eval_bit_ops.go is a new file.
**I didn't modify the other four files, except for** adding a evalXXX() function in every file to classify specific ops.

@hanfei1991 @coocood @shenli @zimulala @tiancaiamao

File: util/types/datum_eval.go [hunk 1]
```diff
@@ -395,3 +395,207 @@ func ComputeIntDiv(a, b Datum) (d Datum, err error) {
 	d.SetInt64(iVal)
 	return d, nil
 }
+
+// decimal2RoundUint convert a MyDecimal to a uint64 after rounding.
```
Comment (hanfei1991) [inline util/types/datum_eval.go]: a uint64 -> an uint64

File: util/types/datum_eval.go [hunk 2]
```diff
@@ -395,3 +395,207 @@ func ComputeIntDiv(a, b Datum) (d Datum, err error) {
...
+
+// ComputeBitXor computes the result of a ^ b.
+func ComputeBitXor(a, b Datum) (d Datum, err error) {
+	aKind, bKind := a.Kind(), b.Kind()
+	if (aKind == KindInt64 || aKind == KindUint64) && (bKind == KindInt64 || bKind == KindUint64) {
+		d.SetUint64(a.GetUint64() ^ b.GetUint64())
+		return
+	}
+
+	// if either is not integer, we round the operands and then use uint64 to calculate.
```
Comment (hanfei1991) [inline util/types/datum_eval.go]: change all if to If

File: util/types/datum_eval.go [hunk 3]
```diff
@@ -395,3 +395,207 @@ func ComputeIntDiv(a, b Datum) (d Datum, err error) {
...
+	uintY, err := decimal2RoundUint(y)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	d.SetUint64(uintX ^ uintY)
+	return d, nil
+}
+
+// ComputeLeftShift computes the result of a >>  b.
```
Comment (hanfei1991) [inline util/types/datum_eval.go]: a useless blank in a >>  b

File: distsql/xeval/eval_bit_ops.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+func (e *Evaluator) evalBitOps(expr *tipb.Expr) (types.Datum, error) {
+	if expr.GetTp() == tipb.ExprType_BitNeg {
+		var result types.Datum
+		if len(expr.Children) != 1 {
+			err := ErrInvalid.Gen("need 1 operands but got %d", len(expr.Children))
```
Comment (hanfei1991) [inline distsql/xeval/eval_bit_ops.go]: 1 operand not operands

**Comment by hanfei1991** [comment]
rest LGTM

File: distsql/xeval/eval.go [hunk 5]
```diff
@@ -54,298 +48,32 @@ type decodedValueList struct {
...
-		return e.evalAnd(expr)
-	case tipb.ExprType_Or:
-		return e.evalOr(expr)
-	case tipb.ExprType_Like:
-		return e.evalLike(expr)
-	case tipb.ExprType_Not:
-		return e.evalNot(expr)
-	case tipb.ExprType_In:
-		return e.evalIn(expr)
+	// data type.
```
Comment (zimulala) [inline distsql/xeval/eval.go]: s/data/Data or remove `.`
The same as line57, line62, line65, line69 and the same as some comments in  plan_test.go

**Comment by XuHuaiyu** [comment]
PTAL
@zimulala @hanfei1991 @coocood

**Comment by XuHuaiyu** [comment]
I'll rewrite this PR after another PRs have been merged.

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @coocood @shenli @zimulala @tiancaiamao 
This commit only involves logical changes.

File: distsql/xeval/eval_bit_ops.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+func (e *Evaluator) evalBitOps(expr *tipb.Expr) (types.Datum, error) {
+	if expr.GetTp() == tipb.ExprType_BitNeg {
+		var result types.Datum
+		if len(expr.Children) != 1 {
+			err := ErrInvalid.Gen("need 1 operand but got %d", len(expr.Children))
```
Comment (shenli) [inline distsql/xeval/eval_bit_ops.go]: Add operator name in the error message.

**Comment by shenli** [comment]
@hanfei1991 PTAL

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
@coocood PTAL

File: distsql/xeval/eval_bit_ops.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		a, err := types.CoerceArithmetic(operand)
+		if err != nil {
+			return result, errors.Trace(err)
+		}
+		return types.ComputeBitNeg(a)
+	}
+	var result types.Datum
```
Comment (shenli) [inline distsql/xeval/eval_bit_ops.go]: Move this to the beginning and remove line 24.

File: util/types/datum_eval.go [hunk 8]
```diff
@@ -395,3 +395,204 @@ func ComputeIntDiv(a, b Datum) (d Datum, err error) {
...
+
+// ComputeBitXor computes the result of a ^ b.
+func ComputeBitXor(a, b Datum) (d Datum, err error) {
+	aKind, bKind := a.Kind(), b.Kind()
+	if (aKind == KindInt64 || aKind == KindUint64) && (bKind == KindInt64 || bKind == KindUint64) {
+		d.SetUint64(a.GetUint64() ^ b.GetUint64())
+		return
+	}
+	// If either is not integer, we round the operands and then use uint64 to calculate.
+	x, err := a.ToDecimal()
```
Comment (shenli) [inline util/types/datum_eval.go]: Use a function to replace the following code block, then we could save a lot of works in ComputeXXX.

**Comment by XuHuaiyu** [comment]
PTAL @shenli

File: util/types/datum_eval.go [hunk 9]
```diff
@@ -395,3 +395,185 @@ func ComputeIntDiv(a, b Datum) (d Datum, err error) {
...
+		d.SetUint64(a.GetUint64() | b.GetUint64())
+		return
+	}
+	// If either is not integer, use decimal to calculate
+	x, err := a.ToDecimal()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	y, err := b.ToDecimal()
```
Comment (shenli) [inline util/types/datum_eval.go]: Use convertNonInt2RoundUint64 instead.

File: util/types/datum_eval.go [hunk 10]
```diff
@@ -395,3 +395,185 @@ func ComputeIntDiv(a, b Datum) (d Datum, err error) {
...
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	uintX, err := decimal2RoundUint(x)
```
Comment (shenli) [inline util/types/datum_eval.go]: convertNonInt2RoundUint64

**Comment by XuHuaiyu** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

**Pull request merged**